### DISCUSSION TITLE
milvus vluster and milvus standalone pending when try to run it with Kubernetes : my-release-milvus-indexnode-954f9678c-qstv9                0/1     Pending     0             93m

### DISCUSSION THREAD

**Discussion by imenselmi** [milvus-io/milvus#35158] [Q&A and General discussion] [answered]
i'm trying to run milvus with gpu and Kubernetes  . I followed the steps in the documentation[link ]( https://milvus.io/docs/install_cluster-helm-gpu.md), also i [Created a K8s cluster with GPU worker nodes](https://milvus.io/docs/prerequisite-gpu.md#How-can-I-start-a-K8s-cluster-with-GPU-worker-nodes) , but the pod always remains pending and is not working.

### 1. Quick Debug Information
* OS/Version : Ubuntu 22.04.4 LTS (Jammy Jellyfish)
* cuda version : 12.2
*  NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2 
*server type : Nvidia L40S : [link](https://www.nvidia.com/en-us/data-center/l40s/)
* Container Runtime Type/Version(e.g. Containerd, CRI-O, Docker): Docker version 27.1.1, build 6312585
* Docker Compose version v2.29.1
* CRI-O version: 1.24.6
* nvidia-container-toolkit  version (1.16.0-1).
* kubectl version :
Client Version: v1.30.3
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.30.0
* minikube version: v1.33.1
* helm Version:"v3.15.3"

## milvus cluster : custom-values.yaml:
` indexNode:
  resources:
    requests:
      nvidia.com/gpu: "1"
    limits:
      nvidia.com/gpu: "1"
  extraEnv:
    - name: CUDA_VISIBLE_DEVICES
      value: "0"
queryNode:
  resources:
    requests:
      nvidia.com/gpu: "1"
    limits:
      nvidia.com/gpu: "1"
  extraEnv:
    - name: CUDA_VISIBLE_DEVICES
      value: "7" `
## milvus standalone : custom-values-standalone.yaml:
` standalone:
  resources:
    requests:
      nvidia.com/gpu: "1"
    limits:
      nvidia.com/gpu: "1"
    extraEnv:
      - name: CUDA_VISIBLE_DEVICES
        value: "4" `

### 3 . Issue:  my-release-milvus-indexnode-954f9678c-qstv9     and standalone-my-release-milvus-standalone-8648c64d5c-cjw6l  are pending : 

` kubectl  get pods
NAME                                                       READY   STATUS      RESTARTS       AGE
gpu-test-5dc7f4d5c7-4sxw8                                  0/1     Pending     0              113m
my-release-etcd-0                                          1/1     Running     0              107m
my-release-etcd-1                                          1/1     Running     0              107m
my-release-etcd-2                                          1/1     Running     0              107m
my-release-milvus-datacoord-6f6c47649-w4s54                1/1     Running     3 (106m ago)   107m
my-release-milvus-datanode-7895b55c66-25hwv                1/1     Running     3 (106m ago)   107m
my-release-milvus-indexcoord-6775954f98-lzjtn              1/1     Running     0              107m
my-release-milvus-indexnode-954f9678c-qstv9                0/1     Pending     0              107m
my-release-milvus-proxy-7889db4d4d-wqxwd                   1/1     Running     3 (106m ago)   107m
my-release-milvus-querycoord-9466bc8f9-rb79h               1/1     Running     3 (106m ago)   107m
my-release-milvus-querynode-669d787c64-vrprj               0/1     Pending     0              107m
my-release-milvus-rootcoord-6d96bb57bf-9xf8s               1/1     Running     3 (106m ago)   107m
my-release-minio-0                                         1/1     Running     0              107m
my-release-minio-1                                         1/1     Running     0              107m
my-release-minio-2                                         1/1     Running     0              107m
my-release-minio-3                                         1/1     Running     0              107m
my-release-pulsar-bookie-0                                 1/1     Running     0              107m
my-release-pulsar-bookie-1                                 1/1     Running     0              107m
my-release-pulsar-bookie-2                                 1/1     Running     0              107m
my-release-pulsar-bookie-init-6qbns                        0/1     Completed   0              107m
my-release-pulsar-broker-0                                 1/1     Running     0              107m
my-release-pulsar-proxy-0                                  1/1     Running     0              107m
my-release-pulsar-pulsar-init-xfmqt                        0/1     Completed   0              107m
my-release-pulsar-recovery-0                               1/1     Running     0              107m
my-release-pulsar-zookeeper-0                              1/1     Running     0              107m
my-release-pulsar-zookeeper-1                              1/1     Running     0              106m
my-release-pulsar-zookeeper-2                              1/1     Running     0              106m
standalone-my-release-etcd-0                               1/1     Running     0              28m
standalone-my-release-milvus-standalone-8648c64d5c-cjw6l   0/1     Pending     0              28m
standalone-my-release-minio-dd7896474-nq5fb                1/1     Running     0              28m `

## nvidia-smi: 
` +---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA L40S                    Off | 00000000:01:00.0 Off |                    0 |
| N/A   36C    P8              33W / 350W |      0MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA L40S                    Off | 00000000:21:00.0 Off |                    0 |
| N/A   35C    P8              32W / 350W |      0MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA L40S                    Off | 00000000:41:00.0 Off |                    0 |
| N/A   34C    P8              32W / 350W |      0MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA L40S                    Off | 00000000:61:00.0 Off |                    0 |
| N/A   32C    P8              32W / 350W |      0MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   4  NVIDIA L40S                    Off | 00000000:81:00.0 Off |                    0 |
| N/A   33C    P8              33W / 350W |      0MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   5  NVIDIA L40S                    Off | 00000000:A1:00.0 Off |                    0 |
| N/A   41C    P0              86W / 350W |  37668MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   6  NVIDIA L40S                    Off | 00000000:C1:00.0 Off |                    0 |
| N/A   45C    P0              86W / 350W |  41988MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   7  NVIDIA L40S                    Off | 00000000:E1:00.0 Off |                    0 |
| N/A   30C    P8              32W / 350W |      0MiB / 46068MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+ `

**Comment by yhmo** [comment]
The compute capability of GPU L40S is 8.9

In our document, the compute capability requirement is: "The compute capability of your GPU device must be one of the following: 6.0, 7.0, 7.5, 8.0, 8.6, 9.0."
https://milvus.io/docs/prerequisite-gpu.md

8.9 is not in the list. This might be the reason why the indexnode/querynode is pending.
Wait @Presburger to confirm.

**Comment by imenselmi** [comment]
@yhmo,  I used DGX, and it works. It's true that this issue is due to the GPU compute capability. Can you support the L40S in the upcoming release? It would be nice if you could add it as a feature.

**Reply by xiaofan-luan** [reply]
@imenselmi 
this problem is we don't have L40 machines.

@liliu-z @Presburger  can you ask help from Nvidia team?

**Reply by presburger** [reply]
@imenselmi Your guess is correct. The currently released Milvus GPU version does not support compute capability 8.9. I will address this issue as soon as possible, or we can provide you with a custom compiled version.

**Reply by imenselmi** [reply, marked as answer]
It now works with the L40S GPU, which has a compute capability of 8.9. I am using the standalone setup for Milvus and have deleted the Milvus cluster. Here is the configuration I used:
standalone:
  resources:
    requests:
      nvidia.com/gpu: "1"
    limits:
      nvidia.com/gpu: "1"
    extraEnv:
      - name: CUDA_VISIBLE_DEVICES
        value: "7"

Additionally, I updated Docker, Kubernetes, and Helm to the latest stable versions. I used the NVIDIA k8s-device-plugin v0.16.1 and with helm chart  : nvdp/nvidia-device-plugin version 0.16.1 instead of v0.15, which resolved the pending issue and allowed Milvus to run with Kubernetes. Please change the link for the installation of the plugin to the official NVIDIA GitHub repository instead of GitLab.

Link : https://github.com/NVIDIA/k8s-device-plugin
